core/trino-main/src/main/java/io/trino/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -349,22 +357,23 @@ public Context(
             this.timeoutInMilliseconds = timeoutInMilliseconds;
             this.session = session;
             this.warningCollector = warningCollector;
+            this.queryStatsCollector = queryStatsCollector;
             this.tableStatsProvider = tableStatsProvider;
         }
 
         public void checkTimeoutNotExhausted()
         {
             if ((NANOSECONDS.toMillis(nanoTime() - startTimeInNanos)) >= timeoutInMilliseconds) {
                 String message = format("The optimizer exhausted the time limit of %d ms", timeoutInMilliseconds);
-                if (ruleStats.isEmpty()) {
+                if (queryStatsCollector.isEmpty()) {
                     message += ": no rules invoked";
                 }
                 else {
                     long timeThresholdNanos = MILLISECONDS.toNanos(timeoutInMilliseconds) / 100;
-                    List<Entry<Rule<?>, RuleInvocationStats>> topRulesByTime = ruleStats.entrySet().stream()
+                    List<QueryRuleStats> topRulesByTime = queryStatsCollector.getQueryRulesStats().stream()
                             // Filter out rules that used less than 1% time that do not change the state
```

Comment:
nit: I would probably just use `getSnapshot()`, or you could also have `getSnapshot(int limit)` as override.
Feel free to skip comment